pkgs/dart_tooling_mcp_server/lib/src/mixins/dart_cli.dart

```diff
@@ -0,0 +1,159 @@
+
+/// Mix this in to any MCPServer to add support for running Dart CLI commands
+/// like `dart fix` and `dart format`.
+///
+/// The MCPServer must already have the [ToolsSupport] and [LoggingSupport]
+/// mixins applied.
+base mixin DartCliSupport on ToolsSupport, LoggingSupport {
+  @override
+  FutureOr<InitializeResult> initialize(InitializeRequest request) {
+    if (request.capabilities.roots == null) {
+      throw StateError(
+        'This server requires the "roots" capability to be implemented.',
+      );
+    }
+    registerTool(dartFixTool, _runDartFixTool);
+    registerTool(dartFormatTool, _runDartFormatTool);
+    return super.initialize(request);
+  }
+
+  /// Implementation of the [dartFixTool].
+  Future<CallToolResult> _runDartFixTool(CallToolRequest request) async {
+    return _runDartCommandInRoots(request, 'dart fix', ['fix', '--apply']);
+  }
+
+  /// Implementation of the [dartFormatTool].
+  Future<CallToolResult> _runDartFormatTool(CallToolRequest request) async {
+    return _runDartCommandInRoots(request, 'dart format', ['format', '.']);
+  }
+
+  /// Helper to run a dart command in multiple project roots.
+  Future<CallToolResult> _runDartCommandInRoots(
+    CallToolRequest request,
+    String commandName,
+    List<String> commandArgs,
+  ) async {
+    final rootConfigs =
+        (request.arguments?['roots'] as List?)?.cast<Map<String, Object?>>();
+    if (rootConfigs == null) {
+      return CallToolResult(
+        content: [TextContent(text: 'Missing required argument `roots`.')],
+        isError: true,
+      );
+    }
+
+    final outputs = <TextContent>[];
+    for (var rootConfig in rootConfigs) {
+      final rootUriString = rootConfig['root'] as String?;
+      if (rootUriString == null) {
+        // This shouldn't happen based on the schema, but handle defensively.
+        return CallToolResult(
+          content: [
+            TextContent(
+              text: 'Invalid root configuration: missing `root` key.',
+            ),
+          ],
+          isError: true,
+        );
+      }
+
+      final rootUri = Uri.parse(rootUriString);
+      if (rootUri.scheme != 'file') {
+        return CallToolResult(
+          content: [
+            TextContent(
+              text:
+                  'Only file scheme uris are allowed for roots, but got '
+                  '$rootUri',
+            ),
+          ],
+          isError: true,
+        );
+      }
+      final projectRoot = Directory(rootUri.toFilePath());
+
+      final result = await Process.run(
+        'dart',
+        commandArgs,
+        workingDirectory: projectRoot.path,
+        runInShell: true,
+      );
+
+      final output = (result.stdout as String).trim();
+      final errors = (result.stderr as String).trim();
+      if (result.exitCode != 0) {
+        return CallToolResult(
+          content: [
+            TextContent(
+              text: '$commandName failed in ${projectRoot.path}:\n$errors',
```

Comment:
Probably should write stdout as well here?

Reply:
Won't writing to standard out mess up the stdio output for the MCP client?

Reply:
I just mean adding the `output` variable above to this text - not writing it to actual stdout.